tusclient/request.py

```diff
@@ -95,14 +105,43 @@ async def perform(self):
         """
         Perform actual request.
         """
-        chunk = self.file.read(self._content_length)
-        self.add_checksum(chunk)
+        if self._upload_checksum:
+            self.add_checksum(self._chunk.reset())
         try:
             async with aiohttp.ClientSession(loop=self.io_loop) as session:
-                async with session.patch(self._url, data=chunk, headers=self._request_headers) as resp:
+                async with session.patch(self._url, data=self._chunk.reset().async_reader(8*1024), headers=self._request_headers) as resp:
                     self.status_code = resp.status
                     self.response_headers = {
                         k.lower(): v for k, v in resp.headers.items()}
                     self.response_content = await resp.content.read()
         except aiohttp.ClientError as error:
             raise TusUploadFailed(error)
+
+
+class ChunkReader(object):
+    def __init__(self, file, start, length):
+        self.file = file
+        self.start = start
+        self.len = length  # to send Content-Length in headers
+        self.remaining = None
+
+    def reset(self):
+        self.file.seek(self.start)
+        self.remaining = self.len
+        return self
+
+    def read(self, size=-1):
+        if self.remaining is None:
+            raise Exception("reset() must be called before first read")
+
+        if size == -1 or size > self.remaining:
+            size = self.remaining
+        data = self.file.read(size)
+        self.remaining -= len(data)
+        return data
```

Comment:
I think it'd be better if we separate the class that mimics the File stream behaviour from every other functionality. This is important so we can easily identify that `read(...)` specifically exists to implement stream readers.

So what I mean is, `async_reader` could just be a stand alone function `async_reader(chunk_reader)`. And `reset` may not be needed if we just used `ChunkReader` directly wherever we need it. So for every place that the ChunkReader is needed, a new instance of it could be created. And then `ChunkReader` itself only contains the `read` method.